source/server/filter_chain_manager_impl.cc

```diff
@@ -335,27 +343,52 @@ void FilterChainManagerImpl::addFilterChainForServerNames(
     const absl::Span<const std::string* const> application_protocols,
+    const std::vector<std::string>& direct_source_ips,
     const envoy::config::listener::v3::FilterChainMatch::ConnectionSourceType source_type,
     const std::vector<std::string>& source_ips,
     const absl::Span<const Protobuf::uint32> source_ports,
     const Network::FilterChainSharedPtr& filter_chain) {
   if (application_protocols.empty()) {
-    addFilterChainForSourceTypes(application_protocols_map[EMPTY_STRING], source_type, source_ips,
-                                 source_ports, filter_chain);
+    addFilterChainForDirectSourceIPs(application_protocols_map[EMPTY_STRING].first,
+                                     direct_source_ips, source_type, source_ips, source_ports,
+                                     filter_chain);
   } else {
     for (const auto& application_protocol_ptr : application_protocols) {
-      addFilterChainForSourceTypes(application_protocols_map[*application_protocol_ptr],
-                                   source_type, source_ips, source_ports, filter_chain);
+      addFilterChainForDirectSourceIPs(application_protocols_map[*application_protocol_ptr].first,
+                                       direct_source_ips, source_type, source_ips, source_ports,
+                                       filter_chain);
+    }
+  }
+}
+
+void FilterChainManagerImpl::addFilterChainForDirectSourceIPs(
+    DirectSourceIPsMap& direct_source_ips_map, const std::vector<std::string>& direct_source_ips,
+    const envoy::config::listener::v3::FilterChainMatch::ConnectionSourceType source_type,
+    const std::vector<std::string>& source_ips,
+    const absl::Span<const Protobuf::uint32> source_ports,
+    const Network::FilterChainSharedPtr& filter_chain) {
+  if (direct_source_ips.empty()) {
+    addFilterChainForSourceTypes(direct_source_ips_map[EMPTY_STRING], source_type, source_ips,
+                                 source_ports, filter_chain);
+  } else {
+    for (const auto& direct_source_ip : direct_source_ips) {
+      addFilterChainForSourceTypes(direct_source_ips_map[direct_source_ip], source_type, source_ips,
+                                   source_ports, filter_chain);
     }
   }
 }
 
 void FilterChainManagerImpl::addFilterChainForSourceTypes(
-    SourceTypesArray& source_types_array,
+    SourceTypesArraySharedPtr& source_types_array_ptr,
     const envoy::config::listener::v3::FilterChainMatch::ConnectionSourceType source_type,
     const std::vector<std::string>& source_ips,
     const absl::Span<const Protobuf::uint32> source_ports,
     const Network::FilterChainSharedPtr& filter_chain) {
+  if (source_types_array_ptr == nullptr) {
+    source_types_array_ptr = std::make_shared<SourceTypesArray>();
+  }
+
+  SourceTypesArray& source_types_array = *source_types_array_ptr;
   if (source_ips.empty()) {
     addFilterChainForSourceIPs(source_types_array[source_type].first, EMPTY_STRING, source_ports,
```

Comment:
Take me a few time to realize this mututes the the upper entry so the allocated SourceTypesArray is not destroyed by end of this function.

This is literally correct.